src/serde_support.rs

```diff
@@ -140,8 +62,23 @@ where
     where
         D: Deserializer<'de>,
     {
-        const FIELDS: &'static [&'static str] = &["capacity", "values"];
-        deserializer.deserialize_struct("CircularQueue", FIELDS, CircularQueueVisitor::new())
+        let mut data = CircularQueueData::deserialize(deserializer)?;
+        // We do not allow the vector to exceed the capacity.
+        if data.values.len() > data.capacity {
+            return Err(Error::invalid_length(
+                data.values.len(),
+                &"serialized vector exceeds capacity of the circular queue",
+            ));
+        }
+        // Grow the vec if needed
+        if data.values.len() < data.capacity {
+            data.values.reserve(data.capacity - data.values.len());
+        }
+        Ok(CircularQueue {
+            data: data.values,
```

Comment:
That's great, don't even have to move the values with this approach... except looking at the docs, looks like both `Vec::reserve` and `Vec::reserve_exact` do not in fact guarantee that capacity will be exactly equal to `vec.len() + additional`. Looks like creating a new vector with `with_capacity()` and moving over the elements is the only approach that will correctly preserve the queue capacity.

This should work:
```rust
let mut queue = CircularQueue::with_capacity(data.capacity);
for x in data.values {
    queue.push(x);
}
Ok(queue)
```

Reply:
I wasn't sure if `CircularQueue` required that the capacity of the backing vector be exact or simply sufficient.  I agree that the `push` is less risky, and removes the error condition as well.

Reply:
No worries on the back and forth.  While I have used/relied on `serde` a lot, I have never had to dig around in the internals.